# <center>Fake-News-Classifier</center>

![Fake-Fact image](https://theleaflet.in/wp-content/uploads/2020/11/190108-fake-thumb.jpg)


# ♦ Introduction ♦

As of Today, It is very hard to find the difference between the fake and the real news as the social media grows wider.Fake news is false or misleading information presented as news. Fake news often has the aim of damaging the reputation of a person or entity, or making money through advertising revenue.To takle this problem i build a machine learning model which can classify the fake news. It may not be the best model but it is a good model.

#About 
## Dataset:
The following dataset used in this project is the famous dataset is taken from the [kaggle](https://www.kaggle.com/datasets/clmentbisaillon/fake-and-real-news-dataset). The dataset contains two csv files named True.csv and Fake.csv which consists of the real news and fake news respectively. Each csv file contains 4 attributes.

**The Attributes of the Dataset**:
1. Title: 
         Title of the News, containing 17903 unique values
2. Text:
         Text Attribute consists of the news written/published in the articles.
3. Subject:
          The Subject Attribute contains the type of news like Political News, World News or just the regular news.
4. Date:
          The date attribute contains the date that this article was posted at.
        


## Project:
The project **Fake News Classifier** is a part of Big Data  Course Project.This Project is completed under Dr.Kanishka Tyagi.In this Project, I made a model analysis so that anyone can choose the best model according to the accuracy of that model to classify the Fake News.In this Project I used  **Bag of Words** method and **Multinomial Naive Bayes** Classifier for the Model Training


## The Model is trained on the basis of 'Text of the Article' in the Dataset

lets install all the required libraries to download the dataset

In [1]:
!pip install pandas opendatasets --upgrade --quiet

In [2]:
import pandas as pd
import opendatasets as opd

## Download 

There are several ways to download the dataset and upload it into your jupyter notebook.
Here is one of the method I am using in this project:
1. Install the [opendatasets](https://github.com/JovianML/opendatasets#opendatasets) library into your Notebook.
2. Copy the URL of the dataset and save the URL into a variable.
3. Use the 'download' function to download the csv file of the dataset.
4. Now you can insert the csv file to your jupyter notebook and perform several operations.

Through this process you may asked to enter the username and API key of Kaggle Account, You need to input those values to proceed Downloading.In case it is already Installed, it does not required

Create an API key in Kaggle.

Step 1:  To do this, go to kaggle.com/ and open your user settings page.
![image1](https://i.stack.imgur.com/jxGQv.png)

Step 2:Next, scroll down to the API access section and click generate to download an API key.
![image2](https://i.stack.imgur.com/Hzlhp.png)
This will download a file called kaggle.json to your computer. You'll use this file in Jupyter Notebook/Google Colab to access Kaggle datasets and competitions.

In [3]:
opd.download('https://www.kaggle.com/datasets/clmentbisaillon/fake-and-real-news-dataset')

Skipping, found downloaded files in ".\fake-and-real-news-dataset" (use force=True to force download)


The dataset thas been downloadeed and extracted

In [4]:
!pip install os --upgrade --quiet
import os
os.listdir('./fake-and-real-news-dataset')

ERROR: Could not find a version that satisfies the requirement os (from versions: none)
ERROR: No matching distribution found for os


['Fake.csv', 'True.csv']

The Dataset containd two CSV files 'True.csv' and 'Fake.csv'

# Data Preperation and Cleaning
Lets Load the File 'Fake.csv' into pandas dataframe and perform some cleaning if required.

In [5]:
fake_df = pd.read_csv('./fake-and-real-news-dataset/Fake.csv',index_col = False)

In [6]:
fake_df.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

**.columns()** method give the list of all the attributes presented in the dataset.

In [7]:
fake_df.columns

Index(['title', 'text', 'subject', 'date'], dtype='object')

**.count()** method provides the Total Number of values present in the Dataset

In [8]:
fake_df.count()

title      23481
text       23481
subject    23481
date       23481
dtype: int64

### Checking if there any NULL values present in the Fake Dataframe

**.info()** provides all the information regarding the attributes in the dataset.

In [9]:
fake_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23481 entries, 0 to 23480
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    23481 non-null  object
 1   text     23481 non-null  object
 2   subject  23481 non-null  object
 3   date     23481 non-null  object
dtypes: object(4)
memory usage: 733.9+ KB


As there are no NULL values in the datset. let's move on to the next step.

In [10]:
fake_df.memory_usage(deep = True)

Index           128
title       5380518
text       61153828
subject     1507065
date        1646426
dtype: int64

As we are only preparing model using 'text' attribute in the dataset, Here seperating the 'text' column from the dataframe and aperform some operations

In [11]:
df_f = fake_df.loc[ : ,['text']]

In [12]:
df_f.head(5)

text
0  Donald Trump just couldn t wish all Americans ...
1  House Intelligence Committee Chairman Devin Nu...
2  On Friday, it was revealed that former Milwauk...
3  On Christmas day, Donald Trump announced that ...
4  Pope Francis used his annual Christmas Day mes...

### Checking for the Duplicate Values in the Fake Dataframe

In [13]:
df_f.count()

text    23481
dtype: int64

Here the Total number of values in the dataset is **23481**

Finding out the total number of unique values in the dataset

In [14]:
n = len(pd.unique(df_f['text']))
print(n)

17455


Here the Total Number of Unique Values are **17455** , Remaining **6026** isn't useful for training the model so let's delete them from the dataset

In [15]:
df_f.drop_duplicates(inplace=True)

In [16]:
df_f.count()

text    17455
dtype: int64

Now, The Duplicate values has been removed from the dataset.

In [17]:
df_f.head(5)

text
0  Donald Trump just couldn t wish all Americans ...
1  House Intelligence Committee Chairman Devin Nu...
2  On Friday, it was revealed that former Milwauk...
3  On Christmas day, Donald Trump announced that ...
4  Pope Francis used his annual Christmas Day mes...

Adding a new column named **TorF** to the dataset to identify wheather the news is Fake or True. For this dataset every value is assigned as 0 as every news is Fake. 

In [18]:
df_f['TorF'] = 0

In [19]:
df_f.head(5)

text  TorF
0  Donald Trump just couldn t wish all Americans ...     0
1  House Intelligence Committee Chairman Devin Nu...     0
2  On Friday, it was revealed that former Milwauk...     0
3  On Christmas day, Donald Trump announced that ...     0
4  Pope Francis used his annual Christmas Day mes...     0

Shuffling the Fake Dataframe

In [20]:
df_f = df_f.sample(frac=1).reset_index(drop=True)

In [21]:
df_f.head(10)

text  TorF
0  Donald Trump has repeatedly defended his blata...     0
1   Every record has been destroyed or falsified,...     0
2  Hillary pandering to the Black Lives Matter cr...     0
3  Saturday Night Live s cold open last night was...     0
4  Project Veritas  newest video from the America...     0
5  If you had any doubt in your mind that Republi...     0
6  President Trump and First Lady Melania Trump m...     0
7  Trump is a millionaire who clearly doesn t nee...     0
8  A customer at a New Orleans Family Dollar stor...     0
9  THE NUMBER OF AMERICANS ON FOOD STAMPS IS AT A...     0

### Performing the Similar Operations on the True News Dataset

Converting **True.csv** into Pandas Dataframe

In [22]:
True_df = pd.read_csv('./fake-and-real-news-dataset/True.csv',index_col = False)

In [23]:
True_df.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017

In [24]:
True_df.columns

Index(['title', 'text', 'subject', 'date'], dtype='object')

In [25]:
True_df.count()

title      21417
text       21417
subject    21417
date       21417
dtype: int64

In [26]:
True_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21417 entries, 0 to 21416
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    21417 non-null  object
 1   text     21417 non-null  object
 2   subject  21417 non-null  object
 3   date     21417 non-null  object
dtypes: object(4)
memory usage: 669.4+ KB


In [27]:
True_df.describe()

title  \
count                                               21417   
unique                                              20826   
top     Factbox: Trump fills top jobs for his administ...   
freq                                                   14   

                                                     text       subject  \
count                                               21417         21417   
unique                                              21192             2   
top     (Reuters) - Highlights for U.S. President Dona...  politicsNews   
freq                                                    8         11272   

                      date  
count                21417  
unique                 716  
top     December 20, 2017   
freq                   182

In [28]:
True_df.memory_usage(deep = True)

Index           128
title       2609581
text       81025844
subject     1447338
date        1578819
dtype: int64

Extracting the **text** attribute from the dataset and perform data cleaning

In [29]:
df_t = True_df.loc[ : ,['text']]

In [30]:
df_t.head(5)

text
0  WASHINGTON (Reuters) - The head of a conservat...
1  WASHINGTON (Reuters) - Transgender people will...
2  WASHINGTON (Reuters) - The special counsel inv...
3  WASHINGTON (Reuters) - Trump campaign adviser ...
4  SEATTLE/WASHINGTON (Reuters) - President Donal...

### Finding out the Total number of duplicate values in the True datsframe

In [31]:
df_t.count()

text    21417
dtype: int64

The Total Number of values in the True Dataframe are **21417**

Finding out the Total Number of Unique Values in the True Dataframe

In [32]:
n = len(pd.unique(df_t['text']))
print(n)

21192


There are **21192** unique values in the True Dataframe. So, there are 225 duplicate values in the dataframe.

Removing the Duplicate Values in the Dataframe

In [33]:
df_t.drop_duplicates(inplace=True)

In [34]:
df_t.count()

text    21192
dtype: int64

In [35]:
df_t.head()

text
0  WASHINGTON (Reuters) - The head of a conservat...
1  WASHINGTON (Reuters) - Transgender people will...
2  WASHINGTON (Reuters) - The special counsel inv...
3  WASHINGTON (Reuters) - Trump campaign adviser ...
4  SEATTLE/WASHINGTON (Reuters) - President Donal...

Adding a new column **TorF** to differentiate the fake news .

In [36]:
df_t['TorF'] = 1

Here all the values are assigned **1** beacuse the present dataset contains all the true news.

In [37]:
df_t.head()

text  TorF
0  WASHINGTON (Reuters) - The head of a conservat...     1
1  WASHINGTON (Reuters) - Transgender people will...     1
2  WASHINGTON (Reuters) - The special counsel inv...     1
3  WASHINGTON (Reuters) - Trump campaign adviser ...     1
4  SEATTLE/WASHINGTON (Reuters) - President Donal...     1

Shuffling the Dataframe

In [38]:
df_t = df_t.sample(frac=1).reset_index(drop=True)

In [39]:
df_t.head(10)

text  TorF
0  SHAMLAPUR, Bangladesh/DHAKA (Reuters) - Myanma...     1
1  MOSCOW (Reuters) - The Paris climate deal is u...     1
2  WASHINGTON (Reuters) - President Donald Trump ...     1
3  BERLIN (Reuters) - Swept into parliament by th...     1
4  NEW YORK (Reuters) - Billionaire activist inve...     1
5  (Reuters) - The continuing controversy over fo...     1
6  WASHINGTON (Reuters) - The U.S. failure to ret...     1
7  SEOUL (Reuters) - U.S. President Donald Trump’...     1
8  DUBLIN (Reuters) - The best way to avoid the r...     1
9  BOSTON (Reuters) - (Editor’s note: This story ...     1

### Combining the two dataframe into one dataframe 

In [40]:
df_combined = pd.concat([df_f,df_t])

In [41]:
df_combined

text  TorF
0      Donald Trump has repeatedly defended his blata...     0
1       Every record has been destroyed or falsified,...     0
2      Hillary pandering to the Black Lives Matter cr...     0
3      Saturday Night Live s cold open last night was...     0
4      Project Veritas  newest video from the America...     0
...                                                  ...   ...
21187  MOSCOW (Reuters) - A Russian media group which...     1
21188  MANILA (Reuters) - Most Filipinos believe only...     1
21189  BEIJING (Reuters) - China is focused on solvin...     1
21190  (Reuters) - The American Civil Liberties Union...     1
21191  WASHINGTON (Reuters) - Former U.S. President B...     1

[38647 rows x 2 columns]

In [42]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38647 entries, 0 to 21191
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    38647 non-null  object
 1   TorF    38647 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 905.8+ KB


Shuffling the Combined Dataframe

In [43]:
df_combined = df_combined.sample(frac=1).reset_index(drop=True)

In [44]:
df_combined.head(10)

text  TorF
0  BEDMINSTER, N.J. (Reuters) - U.S. President Do...     1
1  Russia gave John Podesta 35 millions dollars (...     0
2  Fox Host Bret Baier found himself forced to re...     0
3  Secretary of State Rex Tillerson just shut dow...     0
4  Will Megyn Kelly be working towards a  gotcha ...     0
5  Following the situation in Charlottesville, Tr...     0
6  They re calling them  Y all Qaeda.  It s all a...     0
7  Donald Trump said Monday that Japan s SoftBank...     0
8  Senator Kelly Ayotte (R-NH) is fighting hard t...     0
9  The author of  Clinton Cash  responds to Hilla...     0

# Data pre-processing

The NLP methods using in this Project:
1. Using regular expression, Removing punctuations,special characters and replacing them with spaces abd converting the whole dataframe into lowercase words.
2. Applying Stemming on the whole Dataframe
3. Removing the English Stopwords

The **NLTK** library consists of many useful NLP methods , which we use in this project.


Installing the necessary libraries

In [45]:
import nltk
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

Creating the Object of PorterStemmer() for Stemming.

In [46]:
ps = PorterStemmer()
corpus = []

Iterating each row and applying the regular expressions,removing stopwords and stemming and combining to make a list of sentences.

In [47]:
from  tqdm import tqdm_notebook as tqdm
for i in tqdm(range(0, len(df_combined))):
    review = re.sub('[^a-zA-Z]' , ' ',df_combined['text'][i])
    review = review.lower()
    review = review.split()
    
    review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

C:\Users\rohan\AppData\Local\Temp/ipykernel_14844/3736447135.py:2: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for i in tqdm(range(0, len(df_combined))):


  0%|          | 0/38647 [00:00<?, ?it/s]

Corpus returns a list of all the sentences that are ready to be trained.

In [48]:
corpus

['bedminst n j reuter u presid donald trump thursday declar opioid epidem nation emerg said administr draft paper make offici opioid crisi emerg say offici right emerg trump told report new jersey go draw go make nation emerg seriou problem like never trump spoke week white hous commiss opioid crisi led new jersey governor chri christi recommend presid declar nation emerg declar could help unlock support resourc address drug overdos epidem addit fund expand access variou form treatment give govern flexibl waiv rule restrict expedit action nation emerg typic declar short term crise zika viru outbreak natur disast unclear trump declar mean complex long term public health problem go spend lot time lot effort lot money opioid crisi trump said worldwid problem unit state problem american die drug overdos everi day commiss wrote report note america endur death toll equal septemb th everi three week opioid play role death accord u center diseas control prevent trump held major brief epidem tu

# Bag of Words

The bag-of-words (BOW) model is a representation that turns arbitrary text into fixed-length vectors by counting how many times each word appears. This process is often referred to as vectorization.

## How is BOW useful?
Despite being a relatively basic model, BOW is often used for Natural Language Processing (NLP) tasks like Text Classification. Its strengths lie in its simplicity: it’s inexpensive to compute, and sometimes simpler is better when positioning or contextual info aren’t relevant.

In [49]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 7000)
X = cv.fit_transform(corpus).toarray()

In [50]:
Y = df_combined.iloc[:,1].values

In [51]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y, test_size = 0.20 , random_state = 0)

In [52]:
from sklearn.naive_bayes import MultinomialNB
fake_news_model = MultinomialNB().fit(X_train,Y_train)

In [53]:
Y_pred = fake_news_model.predict(X_test)

In [54]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(Y_test,Y_pred)
matrix

array([[3295,  190],
       [ 218, 4027]], dtype=int64)

In [55]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(Y_test,Y_pred)
accuracy

0.9472186287192755

Here, We got an accuracy of `94.7`%.This may change because using Random Sampling the highest accuracy i got was `95.5`%, Don't Worry it won't be much Different.

## Saving The Model
Saving Machine Learning Model using joblib.

In [59]:
!pip install joblib --upgrade --quiet
import joblib
filename1 = 'BOW_model.sav'
joblib.dump(fake_news_model, filename1)

['BOW_model.sav']

### To load the data:

loaded_model = joblib.load(filename1)